File: app/leetcode/No172.trailingZeroes.js
/**
 * 10只有2和5相乘可以得到，我们只需要找到有多少个2-5对即可
 * 含有2的因子每2个出现1次，含有5的因子每5个出现1次，因此对于每一个因子5一定能找到一个因子2与之对应，我们统计5的因子的个数即可
 * @param {number} n
 * @return {number}
 */
var trailingZeroes = function (n) {

  // let count2 = 0, count5 = 0;
  //
  // for (let i = 2; i <= n; i++) {
  //   let value = i;
  //   while (value % 2 === 0) {
  //     count2++;
  //     value = parseInt(value / 2);
  //   }
  //   while (value % 5 === 0) {
  //     count5++;
  //     value = parseInt(value / 5);
  //   }
  // }
  // return Math.min(count2, count5);

  // let count = 0;
  // for (let i = 1; i <= n; i++) {
  //   let N = i;
  //   while (N) {
  //     if (N % 5 === 0) {
  //       count++;
  //       N = parseInt(N / 5);
  //     } else {
  //       break;
  //     }
  //   }
  // }
  // return count;

  let nCount = 0;
  while (n) {
    nCount += parseInt(n / 5);
    n = parseInt(n / 5);
  }
  return nCount;
};

File: app/leetcode/No172.trailingZeroes.go
package main

func trailingZeroes(n int) int {
	// 给定一个整数 n ，返回 n! 结果中尾随零的数量
	// 5! = 5*4*3*2*1 = 120
	// 6! = 6*5*4*3*2*1 = 720
	// 0 是由 2 的 5相乘产生的
	// 分解所有乘数，看看由多少个 (2,5) 对
	// 转为求 n! 中质因子 2 的个数和 质因子 5 的个数的最小值
	// 由于 质因子 5 的个数一定小于等于质因子 2 的个数：对于质因子 5 ，我们总能找到一个质因子 2 与其对应
	// 所以我们统计质因子 5 的个数即可
	// n/5 + n / (5*5) + n / (5*5*5) + ...
	// 每隔 5 出现 1 次 5
	// 每隔 25 出现 2 次 5
	// 每隔 125 出现 3 次 5
	count := 0
	interval := 5
	for n >= interval {
		count += n / interval
		interval *= 5
	}
	return count
}